File: libs/rs/src/webapi.rs
use crate::error::Error;
use crate::proto::{QueryPlan, QueryResult, QueryResultChunk, StatusCode};
use crate::webapi_bindings::*;
use std::ffi::CStr;
use std::os::raw::c_char;

/// A buffer for WebAPI results
pub struct Buffer<'conn, T> {
    connection: &'conn Connection,
    data_handle: BufferHdl,
    data_size: usize,
    table: Option<T>,
}

impl<'buffer, 'conn: 'buffer, T: 'buffer + flatbuffers::Follow<'buffer, Inner = T>> Buffer<'conn, T> {
    /// Create from data
    pub fn from_data(connection: &'conn Connection, data_handle: BufferHdl, data_size: usize) -> Self {
        Self {
            connection,
            data_handle,
            data_size,
            table: None,
        }
    }

    /// Access a buffer
    pub fn access(&mut self) -> &T {
        match self.table {
            Some(ref t) => t,
            None => unsafe {
                let p = self.connection.access_buffer(self.data_handle);
                let s = std::slice::from_raw_parts(p, self.data_size);
                self.table = Some(flatbuffers::get_root::<T>(s));
                self.table.as_ref().unwrap()
            }
        }
    }
}

impl<'conn, T> Drop for Buffer<'conn, T> {
    fn drop(&mut self) {
        self.connection.release_buffer(self.data_handle);
    }
}

/// A connection to DuckDB
pub struct Connection {
    conn: ConnectionHdl,
}

#[allow(dead_code)]
impl Connection {
    /// Construct a connection from a handle
    pub fn from_handle(conn: ConnectionHdl) -> Self {
        Self { conn }
    }

    /// Disconnect
    pub fn disconnect(&self) -> Result<(), ()> {
        unsafe {
            duckdb_webapi_disconnect(self.conn);
            Ok(())
        }
    }

    /// Access a buffer
    pub(self) fn access_buffer(&self, buffer: BufferHdl) -> *const u8 {
        unsafe { duckdb_webapi_access_buffer(self.conn, buffer) }
    }

    /// Release a buffer
    pub(self) fn release_buffer(&self, buffer: BufferHdl) {
        unsafe { duckdb_webapi_release_buffer(self.conn, buffer) }
    }

    /// Register a buffer
    pub fn register_buffer(&self, buffer_ptr: *const u8, buffer_length: u32) {
        unsafe { duckdb_webapi_register_buffer(self.conn, buffer_ptr, buffer_length) }
    }

    /// Run a query
    pub fn run_query<'conn>(&'conn self, text: *const c_char) -> Result<Buffer<'conn, QueryResult>, Error> {
        let mut r = Response::default();
        unsafe {
            duckdb_webapi_run_query(&mut r, self.conn, text);
            if r.status_code == (StatusCode::ERROR as u64) {
                let err = r.as_error()?;
                return Err(Error::Raw(err));
            }
            let (b, n) = r.as_buffer();
            Ok(Buffer::from_data(self, b, n))
        }
    }

    /// Send a query
    pub fn send_query<'conn>(&'conn self, text: &CStr) -> Result<Buffer<'conn, QueryResult>, Error> {
        let mut r = Response::default();
        unsafe {
            duckdb_webapi_send_query(&mut r, self.conn, text.as_ptr());
            if r.status_code == (StatusCode::ERROR as u64) {
                let err = r.as_error()?;
                return Err(Error::Raw(err));
            }
            let (b, n) = r.as_buffer();
            Ok(Buffer::from_data(self, b, n))
        }
    }

    /// Fetch query results
    pub fn fetch_query_results<'conn>(&'conn self) -> Result<Buffer<'conn, QueryResultChunk>, Error> {
        let mut r = Response::default();
        unsafe {
            duckdb_webapi_fetch_query_results(&mut r, self.conn);
            if r.status_code == (StatusCode::ERROR as u64) {
                let err = r.as_error()?;
                return Err(Error::Raw(err));
            }
            let (b, n) = r.as_buffer();
            Ok(Buffer::from_data(self, b, n))
        }
    }

    /// Analyze query
    pub fn analyze_query<'conn>(&'conn self, text: *const c_char) -> Result<Buffer<'conn, QueryPlan>, Error> {
        let mut r = Response::default();
        unsafe {
            duckdb_webapi_analyze_query(&mut r, self.conn, text);
            if r.status_code == (StatusCode::ERROR as u64) {
                let err = r.as_error()?;
                return Err(Error::Raw(err));
            }
            let (b, n) = r.as_buffer();
            Ok(Buffer::from_data(self, b, n))
        }
    }
}

impl Drop for Connection {
    fn drop(&mut self) {
        unsafe {
            duckdb_webapi_disconnect(self.conn);
        }
    }
}

/// A DuckDB database
pub struct DuckDB {}

#[allow(dead_code)]
impl DuckDB {
    /// Init webapi
    pub fn init() {
        unsafe { duckdb_webapi_init() }
    }

    /// Connect
    pub fn connect() -> Result<Connection, Error> {
        unsafe {
            let conn = duckdb_webapi_connect();
            Ok(Connection::from_handle(conn))
        }
    }
}

File: duckdb/libs/cpp/src/webapi.cc
// Copyright (c) 2020 The DashQL Authors

#include "duckdb_webapi/webapi.h"

#include <cstdio>
#include <memory>
#include <optional>
#include <string_view>
#include <unordered_map>

#include "duckdb.hpp"
#include "duckdb/common/vector_operations/vector_operations.hpp"
#include "duckdb/main/client_context.hpp"
#include "duckdb/parser/parser.hpp"
#include "duckdb/planner/planner.hpp"
#include "duckdb_webapi/codec.h"
#include "duckdb_webapi/json.h"
#include "duckdb_webapi/tablegen.h"
#include "flatbuffers/flatbuffers.h"
#include "spdlog/spdlog.h"

namespace fb = flatbuffers;
using namespace duckdb_webapi;

/// Reset the response
void WebAPI::ContextData::clearRequest() {
    request_status_ = proto::StatusCode::SUCCESS;
    request_error_.reset();
    request_data_ = {nullptr, 0};
}

/// Constructor
WebAPI::ContextData::ContextData()
    : detached_buffers_(), adopted_buffers_(), request_status_(), request_data_({nullptr, 0}), request_error_() {}

/// Register a buffer
std::pair<void*, size_t> WebAPI::ContextData::RegisterBuffer(flatbuffers::DetachedBuffer detached) {
    auto data_ptr = detached.data();
    auto data_size = detached.size();
    detached_buffers_.insert({data_ptr, std::move(detached)});
    return {data_ptr, data_size};
}

/// Register a buffer
std::pair<void*, size_t> WebAPI::ContextData::RegisterBuffer(nonstd::span<std::byte> bytes) {
    auto data_ptr = bytes.data();
    auto data_size = bytes.size();
    adopted_buffers_.insert({data_ptr, AdoptedBuffer{bytes}});
    return {data_ptr, data_size};
}

/// Release a buffer
void WebAPI::ContextData::ReleaseBuffer(void* data) {
    detached_buffers_.erase(data);
    adopted_buffers_.erase(data);
}

/// Constructor
WebAPI::Connection::Connection(std::shared_ptr<duckdb::DuckDB> db)
    : database_(std::move(db)),
      connection_(*database_),
      context_data_(std::make_unique<ContextData>()),
      current_query_id_(),
      current_query_result_() {}

/// Run a SQL query
ExpectedBuffer<proto::QueryResult> WebAPI::Connection::RunQuery(std::string_view text) {
    try {
        // Send the query
        auto result = connection_.SendQuery(std::string{text});
        if (!result->success) return {ErrorCode::QUERY_FAILED, move(result->error)};
        current_query_result_ = move(result);

        // Write the result buffer
        fb::FlatBufferBuilder builder{1024};
        auto query_result_ofs = WriteQueryResult(builder, *current_query_result_, ++current_query_id_, false);
        builder.Finish(query_result_ofs);
        return {builder.Release()};
    } catch (std::exception& e) {
        return {ErrorCode::QUERY_FAILED, e.what()};
    }
}

/// Start a SQL query
ExpectedBuffer<proto::QueryResult> WebAPI::Connection::SendQuery(std::string_view text) {
    try {
        // Send the query
        auto result = connection_.SendQuery(std::string{text});
        if (!result->success) return {ErrorCode::QUERY_FAILED, move(result->error)};
        current_query_result_ = move(result);

        // Write the result buffer
        fb::FlatBufferBuilder builder{1024};
        auto query_result_ofs = WriteQueryResult(builder, *current_query_result_, ++current_query_id_, true);
        builder.Finish(query_result_ofs);
        return {builder.Release()};
    } catch (std::exception& e) {
        return {ErrorCode::QUERY_FAILED, e.what()};
    }
}

/// Fetch query results
ExpectedBuffer<proto::QueryResultChunk> WebAPI::Connection::FetchQueryResults() {
    try {
        // Fetch data if a query is active
        std::unique_ptr<duckdb::DataChunk> chunk;
        nonstd::span<duckdb::LogicalType> types;
        if (current_query_result_ != nullptr) {
            chunk = current_query_result_->Fetch();
            types = current_query_result_->types;
        }
        if (!current_query_result_->success) return {ErrorCode::QUERY_FAILED, move(current_query_result_->error)};

        // Get query result
        fb::FlatBufferBuilder builder{128};
        auto ofs = WriteQueryResultChunk(builder, current_query_id_, chunk.get(), types);
        builder.Finish(ofs);

        // Last chunk?
        if (chunk && chunk->size() == 0) current_query_result_.reset();
        return {builder.Release()};
    } catch (std::exception& e) {
        return {ErrorCode::QUERY_FAILED, e.what()};
    }
}

/// Analyze a SQL query
ExpectedBuffer<proto::QueryPlan> WebAPI::Connection::AnalyzeQuery(std::string_view text) {
    // Parse the statements
    duckdb::Connection conn{*database_};
    duckdb::Parser parser;
    parser.ParseQuery(std::string(text));

    // Begin transaction
    conn.context->transaction.BeginTransaction();

    // Invalid statement count?
    if (parser.statements.size() != 1) return ErrorCode::INVALID_REQUEST;

    // Plan the query
    duckdb::Planner planner{*conn.context};
    planner.CreatePlan(move(*parser.statements.begin()));
    conn.context->transaction.Rollback();

    // Write the plan buffer
    fb::FlatBufferBuilder builder{1024};
    auto plan_ofs = WriteQueryPlan(builder, *planner.plan);

    // Return buffer
    builder.Finish(plan_ofs);
    return {builder.Release()};
}

/// Format a query plan
ExpectedBuffer<proto::FormattedText> WebAPI::Connection::FormatQueryPlan(void* query_plan) {
    auto txt = writeJSON(query_plan, *proto::QueryPlanTypeTable());

    // Encode the query plan
    fb::FlatBufferBuilder builder{txt.size() + 16};
    auto txt_ofs = builder.CreateString(txt);
    auto txt_buf = proto::CreateFormattedText(builder, txt_ofs);

    // Return buffer
    builder.Finish(txt_buf);
    return {builder.Release()};
}

/// Generate a table
ExpectedSignal WebAPI::Connection::GenerateTable(proto::TableSpecification& spec) {
    return duckdb_webapi::generateTable(connection_, spec);
}

/// Constructor
WebAPI::WebAPI() : database_(std::make_shared<duckdb::DuckDB>()), connections_() {}

/// Create a session
WebAPI::Connection& WebAPI::Connect() {
    auto conn = std::make_unique<WebAPI::Connection>(database_);
    auto conn_ptr = conn.get();
    connections_.insert({conn_ptr, move(conn)});
    return *conn_ptr;
}

/// End a session
void WebAPI::Disconnect(Connection* session) { connections_.erase(session); }
